src/analyses/dependence_graph.cpp

```diff
@@ -48,42 +50,54 @@ void dep_graph_domaint::control_dependencies(
 {
   // Better Slicing of Programs with Jumps and Switches
   // Kumar and Horwitz, FASE'02:
-  // Node N is control dependent on node M iff N postdominates one
-  // but not all of M's CFG successors.
+  // "Node N is control dependent on node M iff N postdominates, in
+  // the CFG, one but not all of M's CFG successors."
   //
-  // candidates for M are from and all existing control-depended on
-  // nodes; from is added if it is a goto or assume instruction
-  if(from->is_goto() ||
-     from->is_assume())
-    control_deps.insert(from);
+  // The "successor" above refers to an immediate successor of M.
+  //
+  // When computing the control dependencies of a node N (i.e., "to"
+  // being N), candidates for M are all control statements (gotos or
+  // assumes) from which there is a path in the CFG to N.
+
+  // Add new candidates
+
+  if(from->is_goto() || from->is_assume())
+    control_dep_candidates.insert(from);
+  else if(from->is_end_function())
+  {
+    control_dep_candidates.clear();
+    return;
+  }
+
+  if(control_dep_candidates.empty())
+    return;
+
+  // Get postdominators
 
   const irep_idt id=from->function;
   const cfg_post_dominatorst &pd=dep_graph.cfg_post_dominators().at(id);
 
-  // check all candidates for M
-  for(depst::iterator
-      it=control_deps.begin();
-      it!=control_deps.end();
-      ) // no ++it
-  {
-    depst::iterator next=it;
-    ++next;
+  // Check all candidates
 
-    // check all CFG successors
+  for(const auto &cd : control_dep_candidates)
+  {
+    // check all CFG successors of M
     // special case: assumptions also introduce a control dependency
-    bool post_dom_all=!(*it)->is_assume();
+    bool post_dom_all = !cd->is_assume();
     bool post_dom_one=false;
 
     // we could hard-code assume and goto handling here to improve
     // performance
-    cfg_post_dominatorst::cfgt::entry_mapt::const_iterator e=
-      pd.cfg.entry_map.find(*it);
+    cfg_post_dominatorst::cfgt::entry_mapt::const_iterator e =
+      pd.cfg.entry_map.find(cd);
 
-    assert(e!=pd.cfg.entry_map.end());
+    INVARIANT(
+      e != pd.cfg.entry_map.end(), "cfg has an entry for every location");
```

Comment:
The wording for this invariant is a little confusing - do you mean cfg _should_ have an entry for every location?